Code splitting ready for prime time

markATavail opened the pull request:
Splitting is stable, and has run the Avail tests thousands of times.  There is
a glitch that happens ever few hundred runs related to `Acquire_` on an
Avail mutex, but it's unclear when this problem surfaced.

markATavail committed:
WIP.  Introduced code splitting.

markATavail committed:
Code splitting seems fully functional.
- The only condition (currently) that leads to a code split is
  when an int register holds a value that is consumed downstream.
- As a general explanation, when such a condition (an int register
  existing for some semantic value) exists, and would be destroyed
  by a merge with a path that doesn't have such an int register,
  AND if there is a downstream wish for the value as an int, the
  merge is avoided, and the code between where the int register
  exists and where it's used is duplicated, to avoid losing that
  condition.
- o_VariablesWritten() in FiberDescriptor was passing a weak set
  to setFromCollection(x), which failed because the size wasn't
  stable (items were removed by the garbage collector).  Now it's
  staged into a List.
- Added emitTransformedInstruction() implementation in
  L2_JUMP_IF_UNBOX_INT and L2_UNBOX_INT, to allow improved use of
  int registers after splitting.
- Modified readInt(...) to improve detection of int registers of
  synonyms.  There should be a better technique for detecting
  this kind of situation in general, similar to the search for
  equivalent semantic primitive values.

markATavail committed:
Windows does not support setQuitHandler(..), so we now catch the
exception and ignore it.
- Compare-and-set spin loops now use Thread.onSpinWait().
- Switched to "temurin-17" JVM, since J9 was very slow and didn't
  support profiling on Windows.

markATavail committed:
Merge branch 'development' into mvg_splitter

markATavail committed:
WIP on code splitter.
- L2_BIT_LOGIC_OP now wishes for a pair of int arguments.
- L2_EXTRACT_TAG_ORDINAL's emit...() now improves the case
  that the tag has already been extracted.  It also wishes
  for the (unboxed) tag to exist already in a register.
- Some L2 conditionals now have wishes, but not yet emit
  code to take advantage of the wishes.  The conditionals
  should be reworked to have fewer classes, the same way
  L2_JUMP_IF_COMPARE_INT works.

markATavail committed:
WIP code splitting.
- SSA graphs now check that registers have single definitions.
  This currently fails in the presence of postponed instructions,
  which should be using distinct registers instead of being
  emitted verbatim.

markATavail committed:
Added a one-item cache to AbstractWorkbenchTreeNode, to avoid
having to use the glacial HTML3.2 parser on every tree node
every time a JTree is redrawn or resized.  It's MUCH faster.
- Removed redundant mouseClicked handler (probably a bad git merge).
- Upgraded avail-server's gradle spec to request JVM 17.

markATavail committed:
Merge branch 'mvg_development' into mvg_splitter

markATavail committed:
Code splitting is now stable.
- Added DEAD_END zone type for "error-like" L2 code blocks.
- L2_MOVE and L2_MAKE_IMMUTABLE now regenerate correctly by taking
  the current restriction into account.
- Register and semantic value printing is a bit more compact.

markATavail committed:
Minor tweak, during loading use volatile semantics when
reading shared variables.

markATavail committed:
"Time all tests_time|times" shows the iteration number.
- Edges can now override their labels.
- The binary decision trees produced by multi-way jump instructions
  are now boxed in a yellow Zone.
- Dead-end code (e.g., from a bounds check failure) is now boxed in
  a red Zone.
- Cleaned up and homogenized L2 generation of arithmetic primitives
- Division is now unboxed, if the values can be tested to be ints at
  runtime (numerator ≥ 0 and denominator > 0).  That's because Java's
  treatment of negative values is different from Avail's.

markATavail committed:
Released:
- avail-2.0.0.alpha28
- avail-stdlib-2.0.0.alpha28-1.6.1.alpha17.jar

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/L2Instruction.kt:
```diff
@@ -477,4 +478,7 @@ constructor(
 		}
 	}
+
+	fun interestingConditions(): List<L2SplitCondition> =
```
This really needs to be documented. It's practically criminal that I don't know what's interesting about them. Seems really important to the narrative.

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_CONSTANT.kt:
```diff
@@ -103,4 +106,20 @@ object L2_JUMP_IF_GREATER_THAN_CONSTANT : L2ConditionalJump(
 			ifNotGreater)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OBJECT.kt:
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_GREATER_THAN_OBJECT : L2ConditionalJump(
 			ifNotGreater)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_CONSTANT.kt:
```diff
@@ -104,4 +107,20 @@ object L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_CONSTANT : L2ConditionalJump(
 			ifNotGreaterOrEqual)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_OBJECT.kt:
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_OBJECT : L2ConditionalJump(
 			ifNotGreaterOrEqual)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_CONSTANT.kt:
```diff
@@ -98,4 +101,20 @@ object L2_JUMP_IF_LESS_THAN_CONSTANT : L2ConditionalJump(
 			translator, method, instruction, Opcodes.IFNE, ifLess, ifNotLess)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OBJECT.kt:
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_LESS_THAN_OBJECT : L2ConditionalJump(
 			ifNotLess)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_CONSTANT.kt:
```diff
@@ -104,4 +107,20 @@ object L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_CONSTANT : L2ConditionalJump(
 			ifNotLessOrEqual)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_OBJECT.kt:
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_OBJECT : L2ConditionalJump(
 			ifNotLessOrEqual)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_OBJECTS_EQUAL.kt:
```diff
@@ -112,4 +115,23 @@ object L2_JUMP_IF_OBJECTS_EQUAL : L2ConditionalJump(
 			translator, method, instruction, Opcodes.IFNE, ifEqual, ifNotEqual)
 	}
+
+	//TODO override emit...
```
Is this `TODO` concerning?

markATavail replied:
This and some related comments have been obviated by completely rewriting the conditional branch instruction hierarchy, specifically introducing NumericCompare as a helper.

toddATavail commented on avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_UNBOX_INT.kt:
```diff
@@ -126,4 +131,70 @@ object L2_JUMP_IF_UNBOX_INT : L2ConditionalJump(
+		val generator = regenerator.targetGenerator
+		val manifest = generator.currentManifest
+		val restriction = manifest.restrictionFor(source.semanticValue())
+		// See if there's an int version of a synonym of the source.  There
+		// must be a less messy way of doing this.
+		val sourceInt = manifest.semanticValueToSynonym(source.semanticValue())
+			.semanticValues()
+			.map(::L2SemanticUnboxedInt)
+			.firstOrNull(manifest::hasSemanticValue)
+			?: L2SemanticUnboxedInt(source.semanticValue())
```
Based on the comment, maybe this needs to be factored into a named operation? Or is this the only place where you need to do this?

toddATavail commented on avail/src/main/kotlin/avail/optimizer/L2Optimizer.kt:
```diff
@@ -222,6 +224,194 @@ class L2Optimizer internal constructor(
+					}
+				}
+			}
+		}
```
This looks like it should be removed or conditionalized in some way.

toddATavail reviewed (changes requested):
I had small quibbles with your comments (or lack thereof) occasionally, and questions about some TODOs, but your techniques and logic look solid. Glad we have tests to prove general correctness, though! Would be extraordinarily hard by eye!

markATavail committed:
When the debugJVM ("Debuug JVM translation" menu item) is enabled,
a .l2 file is written (along with .l1, .dot, and .class).  The
line number information in the .class file now references the .l2
lines instead of the L2 offset.  The referenced source file now
uses the local .l2 name, but IntelliJ doesn't present it correctly
yet.

markATavail committed:
Code splitting is implemented.
- When the debugJVM ("Debug JVM translation" menu item) is enabled,
a .l2 file is written (along with .l1, .dot, and .class).  The
line number information in the .class file now references the .l2
lines instead of the L2 offset.  The referenced source file now
uses the local .l2 name, but IntelliJ doesn't present it correctly
yet.
- Added generic parameter `<K: RegisterKind<K>>` in many places.
- Tweaked date/time literals in tests to use milliseconds instead
  of accidentally using centiseconds.
- CFG now implements forwardVisit{} and backwardVisit{}.
- L2_MULTIWAY_JUMP is now parameterized by a usage-specific
  "splitter" for tags, variants, etc.
- Code splitter knows to ask operations and primitives for
  proposed L2SplitConditions, and also has heuristics for cold
  paths to avoid over-splitting when not profitable.
- Added L2GeneratorInterface above L2Generator and L2Regenerator.
- Branches that have edges with impossible constraints (bottom)
  are eliminated.
- A concurrency bug remains in Avail semaphores, where an
  acquire or release fails once every few thousand runs of all
  tests.
- NumericComparator now has a factorization of the different
  comparison operations, irrespective of whether it's for boxed,
  ints, or doubles, and whether it's against a constant.
- Mutually invariant relationships between semantic values (e.g.,
  x <-> Int(x), x <-> Int(Tag(x)), or some primitives)
  automatically propagate type strengthening within the manifest.
- Immutability analysis happens near the end of optimization,
  rather than being respected during earlier passes.
- Cold blocks are shown in cyan, and warm edges are shown as
  lines that are 5 pixels wide.
- Manifests on edges show entries added since the previous node's
  inputs as green, and any alterations as yellow.
- Renamed "translateToLevelTwo" to have a yellow flower Unicode
  character at the start, so that it's easier to locate the frame
  in the debugger that should be restarted to re-do optimization.

markATavail committed:
P_MethodFromName previously added the CanFold flag, which broke
deserialization.  It automatically introduced a semantic restriction
that strengthened the call site to be an instance type on the
looked up method, and that instance type became a literal of the
calling code.  Deserialization of a method defined in the current
module was not yet supported, so I fixed it.
- Also added a horizontal scroll bar in CodePane, specifically
  because wide strings without breakable whitespace made
  horizontal scrolling pretty awkward in Swing.

markATavail committed:
- Transcript writes now clear the undo stack.  This was leaking
  references, leading to out-of-memory failures.
- A one-time-ever bug showed up in the "fork and join with result"
  test.  The main fiber tested a forked fiber to see if it was
  done, then read its result, but exitFiber() wrote these two
  volatiles in the reverse order.  Fixed, and ran the entire
  concurrency test suite >900,000 times without problem.

markATavail committed:
Added L2_ENTER_L2_CHUNK_FOR_CALL, which captures arguments,
and removed L2_GET_ARGUMENT.  Also eliminated hidden state
annotation flag CURRENT_ARGUMENTS.
- (re-)Added L2WriteBoxedVectorOperand.
- This is precursor work for changing the calling
  convention to be more efficient and direct – i.e., to get
  rid of Interpreter.argsBuffer.

markATavail committed:
Made `L2Instruction.operation` private, in preparation
for eliminating L2Operation and reworking L2Instruction
to directly contain the variation.  Operands will occur
in fields directly, rather than an array.
- Next step will be to support both instruction kinds
  at the same time, then migrate the existing operations
  into instruction variations, then remove L2Operation,
  and then revisit these shims in the context of the new
  paradigm.

markATavail committed:
Both L2OldInstruction (using L2Operation) and subclasses of
L2NewInstruction can coexist during migration to the latter.
L2_UNBOX_INT has been switched from an L2Operation subclass
to an L2NewInstruction subclass, and all tests run.
- Moved `unsupportedOperation()` and `unsupported` into top
  level fun/val in AvailUnsupportedOperationException, and
  made unsupported be an Any extension property, so it can
  be used directly anywhere, making it look like a keyword.

markATavail committed:
Translated L2_JUMP_IF_UNBOX_INT from subclass of
L2OldInstruction/L2Operation into L2NewInstruction subclass.
- Split L2ConditionalJump into L2[Old|New]ConditionalJump.
- Split L2ControlFlowOperation into L2[Old|New]ControlFlowOperation.

markATavail committed:
Translated L2_TYPE_UNION from subclass of
L2OldInstruction/L2Operation into L2NewInstruction subclass.
- P_TypeUnion now generates this, as well as performing
  union identities and declaring split conditions that expose
  those identities.
- topMeta and anyMeta are now exposed as properties rather
  than functions.

markATavail committed:
- L2_MOVE/L2_PHI (and [_BOXED|_INT|_FLOAT]) is now under
  L2NewInstruction.
- L2_STRIP_MANIFEST no longer requires separate moves – it now has
  a vector of inputs and a vector of corresponding outputs, and
  its instructionWasAdded() clears other semantic values and
  registers.  Reworked P_RestartContinuation[WithArguments] to
  accommodate those changes.  The register colorer now attempts
  to eliminate these corresponding moves when possible, as it
  does with L2_MOVE.
- Fields in L2NewInstruction subclasses preserve their order
  when the instruction is being described.
- Blocks with phi instructions now show their incoming edges
  annotated as e.g., "#1/#3".
- Corrected some problems with extractTupleElement().
- Introduced OperandTypeMap to tweak initialization order of
  L2Operand classes and L2NamedOperandType.
- During the postponement pass, if an instruction with no
  side effect produces an output that's "alwaysIn" in all
  successor edges, don't postpone it, since it will always be
  needed, and it would lead to unnecessary duplication of the
  instruction.
- REMOVE_DEAD_CODE_AFTER_PHI_MOVES now only pays attention to
  the registers, not the semantic values.

markATavail committed:
- L2_STRIP_MANIFEST is now under L2NewInstruction.
- Added safety check for accidental val fields holding an
  L2Operand in an L2NewInstruction subclass.

markATavail committed:
Migrated operation classes to L2NewInstruction:
- L2_BOX_FLOAT
- L2_BOX_INT
- L2_UNBOX_FLOAT

- Removed never-used L2_FLOAT_TO_INT.
- Improved diagnostic for accidental use of val operand field.

markATavail committed:
Migrated L2_ADD_INT_TO_INT.
- Tidied jump/jumpOrFallThrough

markATavail committed:
Migrated L2_SUBTRACT_INT_MINUS_INT, L2_MULTIPLY_INT_BY_INT.kt,
and (unused) L2_DIVIDE_INT_BY_INT.

markATavail committed:
Migrated operations used by the L1 interpreter:
- L2_DECREMENT_COUNTER_AND_REOPTIMIZE_ON_ZERO
- L2_INTERPRET_LEVEL_ONE
- L2_JUMP
- L2_JUMP_BACK
- L2_PREPARE_NEW_FRAME_FOR_L1
- L2_REENTER_L1_CHUNK_FROM_CALL
- L2_REENTER_L1_CHUNK_FROM_INTERRUPT
- L2_TRY_OPTIONAL_PRIMITIVE
- L2_UNREACHABLE_CODE
- L2_VIRTUAL_CREATE_LABEL

markATavail committed:
Migrated L2_MOVE_CONSTANT[_BOXED|_INT|_FLOAT].
- Adjusted L2_MOVE subclasses to be defined inside
  the body of the L2_MOVE class, rather than file level.

markATavail committed:
Migrated L2_CREATE_FUNCTION and L2_MOVE_OUTER_VARIABLE.

markATavail committed:
Migrated continuation-themed L2 operations.
- InstructionLayout.operandFields is now in declaration
  order (although JVM reflection says the order is
  unspecified)
- The regenerator now automatically does a basic
  transformation on each instruction, which makes it use
  the new graph's registers and blocks instead of the
  original graph.

markATavail committed:
Migrated L2OldConditionalJump hierarchy into L2NewConditionalJump.

markATavail committed:
Tweaked printing of branch instructions that use NumericComparator.

markATavail committed:
Migrated rest of L2OldControlFlowOperation to L2NewControlFlowInstruction.

markATavail committed:
Migrated more call/return operations from L2OldControlFlowOperation to L2NewControlFlowInstruction.
- Also variable get/set and some hooks.
- Fixed some unrelated Boolean val properties that forgot a `get()`.
- Operand fields can now be inherited in L2NewInstruction hierarchy.

markATavail committed:
Migrated the remaining subclasses from L2Operation to L2NewInstruction.
- Removed L2Operation and other L2Old* classes.
- Collapsed L2Instruction and L2NewInstruction.
- Renamed various L2New* -> L2*.

markATavail committed:
WIP all tests pass, ready to merge to development.
- An L2 pass replaces constant-valued registers with ones having no definition.
- Replace-NAry primitives now use new recursivelyUpdate().
- TreeTupleDescriptor.o_TupleAtPuttingCanDestroy() was failing to account for the
  preToggle when updating the hash.
- Corrected some light-mode code.
- Introduced ChainedComparator, thenBy(), and compareChained() for multi-level sort.
- Replaced java.lang.long.bitCount() with Kotlin's countOneBits().
- L2ArbitraryConstantOperand<*> now subsumes L2PrimitiveOperand and
  L2SelectorOperand.
- Added L2_APPEND_TO_TUPLE, used by P_TupleAppend and when the optimizer detects
  that a concatenate has a single-element tuple on the right.
- BitOperations now have reduction rules, although they're currently unused.

toddATavail reviewed (approved):
There's a lot here, but I glanced at most everything. I will have to take your word (and test results) that everything is ready. We ought to fix the 2x performance regression, though; it certainly wasn't the goal for code splitting to be a _de_-optimization.
